File: packages/edge/src/middleware/clientAuth.ts
import { Request, Response, NextFunction } from 'express';
import logger from '../config/logger';
import { tokenMirrorService } from '../services/tokenMirrorService';
import { tokenUsageTracker } from '../services/tokenUsageTracker';

// Unsecured tokens for testing purposes
const UNSECURED_CLIENT_TOKEN = 'gatrix-unsecured-client-api-token';
const UNSECURED_SERVER_TOKEN = 'gatrix-unsecured-server-api-token';
const UNSECURED_EDGE_TOKEN = 'gatrix-unsecured-edge-api-token';

export interface ClientRequest extends Request {
  clientContext?: {
    apiToken: string;
    applicationName: string;
    /**
     * Environment identifier (environmentName value).
     * This is the standard external identifier for environments.
     */
    environment: string;
    clientVersion?: string;
    platform?: string;
    tokenName?: string;
  };
}

/**
 * Client authentication middleware
 * Validates required headers and API token from client requests
 * Uses locally mirrored tokens for validation (no backend call needed)
 *
 * Environment is extracted from URL path parameter (:environment)
 * instead of x-environment header for cleaner API design.
 */
export function clientAuth(req: ClientRequest, res: Response, next: NextFunction): void {
  // Extract token from multiple sources
  const authHeader = req.headers['authorization'] as string | undefined;
  let apiToken = req.headers['x-api-token'] as string | undefined;

  if (!apiToken && authHeader?.startsWith('Bearer ')) {
    apiToken = authHeader.substring(7);
  }

  // Support token in query for testing convenience
  if (!apiToken) {
    apiToken = (req.query.token as string) || (req.query.apiToken as string);
  }

  // Support applicationName in query
  const applicationName =
    (req.headers['x-application-name'] as string) ||
    (req.query.appName as string) ||
    (req.query.applicationName as string);

  // Get environment from URL path parameter instead of header
  const environment = req.params.environment as string;
  const clientVersion = req.headers['x-client-version'] as string | undefined;
  const platform = req.headers['x-platform'] as string | undefined;

  // Validate required parameters
  if (!apiToken) {
    logger.warn('Missing API token in client request', { url: req.originalUrl, query: req.query });
    res.status(401).json({
      success: false,
      error: {
        code: 'MISSING_API_TOKEN',
        message: 'x-api-token header or token query parameter is required',
      },
    });
    return;
  }

  if (!applicationName) {
    logger.warn('Missing application name in client request', { url: req.originalUrl });
    res.status(401).json({
      success: false,
      error: {
        code: 'MISSING_APPLICATION_NAME',
        message: 'x-application-name header or appName query parameter is required',
      },
    });
    return;
  }

  // Handle Unsecured Tokens (Bypass)
  if (
    apiToken === UNSECURED_CLIENT_TOKEN ||
    apiToken === UNSECURED_SERVER_TOKEN ||
    apiToken === UNSECURED_EDGE_TOKEN
  ) {
    req.clientContext = {
      apiToken,
      applicationName,
      environment,
      clientVersion,
      platform,
      tokenName: 'Unsecured Testing Token',
    };
    logger.debug('Authenticated with unsecured token', { token: apiToken, environment });
    return next();
  }

  // Validate environment from path parameter
  if (!environment) {
    res.status(400).json({
      success: false,
      error: {
        code: 'MISSING_ENVIRONMENT',
        message: 'Environment is required in URL path (e.g., /api/v1/client/{environment}/...)',
      },
    });
    return;
  }

  // Validate API token using mirrored tokens
  const validation = tokenMirrorService.validateToken(apiToken, 'client', environment);

  if (!validation.valid) {
    const errorMessages: Record<string, { code: string; message: string }> = {
      not_found: { code: 'INVALID_TOKEN', message: 'Invalid API token' },
      expired: { code: 'TOKEN_EXPIRED', message: 'API token has expired' },
      invalid_type: {
        code: 'INVALID_TOKEN_TYPE',
        message: 'Token is not authorized for client API access',
      },
      invalid_environment: {
        code: 'INVALID_ENVIRONMENT',
        message: 'Token is not authorized for this environment',
      },
    };

    const error = errorMessages[validation.reason || 'not_found'];

    logger.warn('Client authentication failed', {
      reason: validation.reason,
      environment,
      applicationName,
    });

    res.status(401).json({
      success: false,
      error,
    });
    return;
  }

  // Record token usage for tracking (skip unsecured tokens with id=0)
  if (validation.token?.id && validation.token.id > 0) {
    tokenUsageTracker.recordUsage(validation.token.id);
  }

  // Set client context
  req.clientContext = {
    apiToken,
    applicationName,
    environment,
    clientVersion,
    platform,
    tokenName: validation.token?.tokenName,
  };

  logger.debug('Client authenticated', {
    applicationName,
    environment,
    clientVersion,
    platform,
    tokenName: validation.token?.tokenName,
  });

  next();
}

File: packages/edge-rust/src/middleware/client_auth.rs
// Gatrix Edge Rust - Client Authentication
// Validates client API tokens and resolves environment context

use std::sync::Arc;
use actix_web::{HttpRequest, HttpResponse};
use log::{debug, warn};
use regex::Regex;
use std::sync::LazyLock;

use crate::services::token_mirror::TokenMirrorService;
use crate::services::environment_registry::EnvironmentRegistry;

/// Unsecured token regex: unsecured-{org}:{project}:{env}-{type}-api-token
static UNSECURED_TOKEN_REGEX: LazyLock<Regex> = LazyLock::new(|| {
  Regex::new(r"^unsecured-([^:]+):([^:]+):(.+)-(server|client|edge)-api-token$").unwrap()
});

/// Legacy unsecured tokens
const LEGACY_TOKENS: &[&str] = &[
  "unsecured-client-api-token",
  "unsecured-server-api-token",
  "unsecured-edge-api-token",
];

const LEGACY_ENV_NAME: &str = "development";

/// Client context extracted from authentication
#[derive(Debug, Clone)]
pub struct ClientContext {
  pub api_token: String,
  pub application_name: String,
  pub environment_id: String,
  pub cache_key: String,
  pub client_version: Option<String>,
  pub platform: Option<String>,
  pub token_name: Option<String>,
}

/// Extract API token from request headers/query
fn extract_api_token(req: &HttpRequest) -> Option<String> {
  // Try x-api-token header
  if let Some(token) = req.headers().get("x-api-token") {
    if let Ok(t) = token.to_str() {
      return Some(t.to_string());
    }
  }

  // Try Authorization: Bearer header
  if let Some(auth) = req.headers().get("authorization") {
    if let Ok(a) = auth.to_str() {
      if let Some(token) = a.strip_prefix("Bearer ") {
        return Some(token.to_string());
      }
    }
  }

  // Try query parameters
  let query = actix_web::web::Query::<std::collections::HashMap<String, String>>::from_query(
    req.query_string(),
  );
  if let Ok(params) = query {
    if let Some(token) = params.get("token").or(params.get("apiToken")) {
      return Some(token.clone());
    }
  }

  None
}

/// Extract application name from request
fn extract_application_name(req: &HttpRequest) -> Option<String> {
  if let Some(name) = req.headers().get("x-application-name") {
    if let Ok(n) = name.to_str() {
      return Some(n.to_string());
    }
  }

  let query = actix_web::web::Query::<std::collections::HashMap<String, String>>::from_query(
    req.query_string(),
  );
  if let Ok(params) = query {
    if let Some(name) = params.get("appName").or(params.get("applicationName")) {
      return Some(name.clone());
    }
  }

  None
}

/// Authenticate a client request and return ClientContext
pub async fn authenticate_client(
  req: &HttpRequest,
  token_mirror: &Arc<TokenMirrorService>,
  env_registry: &Arc<EnvironmentRegistry>,
) -> Result<ClientContext, HttpResponse> {
  let api_token = extract_api_token(req).ok_or_else(|| {
    warn!("Missing API token in client request");
    HttpResponse::Unauthorized().json(serde_json::json!({
      "success": false,
      "error": {
        "code": "MISSING_API_TOKEN",
        "message": "x-api-token header or token query parameter is required"
      }
    }))
  })?;

  let application_name = extract_application_name(req).ok_or_else(|| {
    warn!("Missing application name in client request");
    HttpResponse::Unauthorized().json(serde_json::json!({
      "success": false,
      "error": {
        "code": "MISSING_APPLICATION_NAME",
        "message": "x-application-name header or appName query parameter is required"
      }
    }))
  })?;

  let client_version = req
    .headers()
    .get("x-client-version")
    .and_then(|v| v.to_str().ok())
    .map(|s| s.to_string());
  let platform = req
    .headers()
    .get("x-platform")
    .and_then(|v| v.to_str().ok())
    .map(|s| s.to_string());

  // 1. Try unsecured token format
  if let Some(captures) = UNSECURED_TOKEN_REGEX.captures(&api_token.clone()) {
    let env_id = captures.get(3).unwrap().as_str();
    let environment_id = env_registry
      .resolve_environment_id(env_id)
      .await
      .unwrap_or_else(|| env_id.to_string());

    debug!("Authenticated with unsecured format token, env={}", environment_id);

    return Ok(ClientContext {
      api_token,
      application_name,
      environment_id: environment_id.clone(),
      cache_key: environment_id,
      client_version,
      platform,
      token_name: Some(format!("Unsecured Token ({})", env_id)),
    });
  }

  // 2. Legacy unsecured tokens
  if LEGACY_TOKENS.contains(&api_token.as_str()) {
    let env_id = env_registry
      .resolve_environment_id(LEGACY_ENV_NAME)
      .await
      .ok_or_else(|| {
        HttpResponse::Unauthorized().json(serde_json::json!({
          "success": false,
          "error": {
            "code": "ENVIRONMENT_NOT_FOUND",
            "message": "Could not resolve environment for legacy token"
          }
        }))
      })?;

    debug!("Authenticated with legacy unsecured token, env={}", env_id);

    return Ok(ClientContext {
      api_token,
      application_name,
      environment_id: env_id.clone(),
      cache_key: env_id,
      client_version,
      platform,
      token_name: Some("Legacy Unsecured Token".to_string()),
    });
  }

  // 3. Real production token
  let validation = token_mirror.validate_token(&api_token, "client").await;

  if !validation.valid {
    let reason = validation.reason.as_deref().unwrap_or("not_found");
    let (code, message) = match reason {
      "expired" => ("TOKEN_EXPIRED", "API token has expired"),
      "invalid_type" => (
        "INVALID_TOKEN_TYPE",
        "Token is not authorized for client API access",
      ),
      "invalid_environment" => (
        "INVALID_ENVIRONMENT",
        "Token is not authorized for this environment",
      ),
      _ => ("INVALID_TOKEN", "Invalid API token"),
    };

    warn!("Client authentication failed: {}", reason);

    return Err(HttpResponse::Unauthorized().json(serde_json::json!({
      "success": false,
      "error": { "code": code, "message": message }
    })));
  }

  let token = validation.token.as_ref();
  let token_env_id = token
    .and_then(|t| t.environment_id.as_ref())
    .ok_or_else(|| {
      HttpResponse::Unauthorized().json(serde_json::json!({
        "success": false,
        "error": {
          "code": "INVALID_TOKEN",
          "message": "Token does not have a specific environment binding"
        }
      }))
    })?;

  let env_id = env_registry
    .resolve_environment_id(token_env_id)
    .await
    .unwrap_or_else(|| token_env_id.clone());

  if !env_registry.has_environment(&env_id).await {
    return Err(HttpResponse::Unauthorized().json(serde_json::json!({
      "success": false,
      "error": {
        "code": "ENVIRONMENT_NOT_FOUND",
        "message": format!("Could not resolve environment: {}", token_env_id)
      }
    })));
  }

  debug!("Client authenticated, env={}", env_id);

  Ok(ClientContext {
    api_token,
    application_name,
    environment_id: env_id.clone(),
    cache_key: env_id,
    client_version,
    platform,
    token_name: token.map(|t| t.token_name.clone()),
  })
}
